File: algorithms/javaalgo/src/test/java/leetcode/number1to100/number51to60/number55/JumpGameTest.java
package leetcode.number1to100.number51to60.number55;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class JumpGameTest {

    @Test
    void test() {
        int[] nums1 = {2, 3, 1, 1, 4};
        int[] nums2 = {3, 2, 1, 0, 4};

        assertThat(canJump(nums1)).isTrue();
        assertThat(canJump(nums2)).isFalse();
    }

    public boolean canJump(int[] nums) {
        int length = nums.length;
        int lastJump = 0;

        for (int i = 0; i < length; i++) {
            if (i > lastJump) {
                return false;
            }
            lastJump = Math.max(lastJump, i + nums[i]);
        }
        return true;
    }
}

File: algorithms/javascriptalgo/leetcode/number51-60/number55/jumpGame.test.js
test("test", () => {
  const nums1 = [2, 3, 1, 1, 4];
  const nums2 = [3, 2, 1, 0, 4];

  expect(canJump(nums1)).toBeTruthy();
  expect(canJump(nums2)).toBeFalsy();
});

// 최대 점프 길이를 의미함
function canJump(nums) {
  let lastJump = 0;
  for (let i = 0; i < nums.length; i++) {
    if (lastJump < i) {
      return false;
    }
    lastJump = Math.max(lastJump, i + nums[i]);
  }
  return true;
}
